### ISSUE TITLE
Lenient validation and merging for non-strict `listType=map`

### ISSUE THREAD

**Issue by apelisse** [kubernetes-sigs/structured-merge-diff#234]
Right now, we have some lists that are marked as `listType=map` that are not strictly maps in Kubernetes. In that case, everything goes wrong in smd because the list is invalid (and can't be handled), so it's fully rejected. We can't even track the fields, which causes us to fail and return an error which is ignored by Kubernetes, and the managed fields are entirely reset.

Unfortunately, it means that objects that have such a list can't be server-side applied at all. We think it'd be simpler to merely ignore these (fairly rare) occurences and try to do the best thing possibe rather than fail.

This means:
1. When we validate an object with such a list, we should probably not fail altogether (though we should probably reject a request to server-side apply an object like this).
1. When tracking the fields, we should give duplicate fields to the same owner.
1. When server-side applying without the field, we need to be careful about how the field should be removed.
1. When we merge a list that has such a field, we need to be careful, especially if the server-side apply contains the field (should we remove both other occurences? Probably).

**Comment by apelisse** [comment]
cc @thockin @jpbetz

**Comment by jpbetz** [comment]
Do we have the list of fields that fall into this "marked as listType=map but not validated as having unique keys" category?

**Comment by apelisse** [comment]
Not really but, that's a great question:
1. I'm pretty sure  CRDs don't allow that and never will, so there's no chance people start abusing that mechanism there (IOW, actual validation matches SSA validation)
2. Since it's only in built-ins, and because a lot of these were not designed for that initially (unfortunately), we have:
environment variables and a few containerports.

I considered adding a new, separate tag, but that would require a KEP and make things impossible to backport. I think we need to make it clear that this is a work-around and shouldn't be used for any new list (again, only for built-ins so we have a lot more control over that).

**Comment by thockin** [comment]
There are two distinct failure modes, the one we have usually run into is client-side apply vs. `Service.spec.ports`, where the keys are looser than validation, and SSA is the SOLUTION (the keys match validation).

The more recent one is where SSA's keys are more strict than validation.  So far we've found:
* `Pod.spec.containers[].ports`
* `Pod.spec.containers[].env`
* `Pod.spec.containers[].envFrom`

I admit that I haven't scoured the API for this case, and I don't honestly know how I would in any reasonable amount of time. :(  IOW, there's ikely to be more

**Comment by jpbetz** [comment]
I did a scan. Here's what I found:

- [StorageVersionStatus.StorageVersions](https://github.com/kubernetes/kubernetes/blob/5a5ebfd88b2899f2116dc858100832f30ac33cf3/pkg/apis/apiserverinternal/validation/validation.go#L70-L72) - listMapKey=apiServerID - validation missing, Alpha API (we should fix now!)
- [Container.Ports](https://github.com/kubernetes/kubernetes/blob/5a5ebfd88b2899f2116dc858100832f30ac33cf3/pkg/apis/core/validation/validation.go#L3305C2-L3319) - validates unique name, listMapKey=containerPort, protocol
- [PodSpec.TopologySpreadConstraints](https://github.com/kubernetes/kubernetes/blob/c5100c0d11fb32930dc49fd0a6e2ec3a5aa99d9e/pkg/apis/core/validation/validation.go#LL6940C6-L6940C39) - validation missing, listMapKey=whenUnsatisfiable
- [EphemeralContainerCommon.Ports](https://github.com/kubernetes/kubernetes/blob/c5100c0d11fb32930dc49fd0a6e2ec3a5aa99d9e/pkg/apis/core/validation/validation.go#L3097-L3098) - same as Container.Ports
- [ServiceSpec.Ports](https://github.com/kubernetes/kubernetes/blob/c5100c0d11fb32930dc49fd0a6e2ec3a5aa99d9e/pkg/apis/core/validation/validation.go#L4977-L4982) - same as ContainerPorts (except keys are port, protocol)
- [Container.Env](https://github.com/kubernetes/kubernetes/blob/5a5ebfd88b2899f2116dc858100832f30ac33cf3/pkg/apis/core/validation/validation.go#L2387) - duplicates explicitly allowed, listMapKey=name
- [PodSpec.ImagePullSecretes](https://github.com/kubernetes/kubernetes/blob/5a5ebfd88b2899f2116dc858100832f30ac33cf3/pkg/apis/core/validation/validation.go#L3516) - validation missing, listMapKey=name
- [PodSpec.HostAliases](https://github.com/kubernetes/kubernetes/blob/5a5ebfd88b2899f2116dc858100832f30ac33cf3/pkg/apis/core/validation/validation.go#L3616) - validation missing, listMapKey=ip
- [ServiceAccount.Secrets](https://github.com/kubernetes/kubernetes/blob/5a5ebfd88b2899f2116dc858100832f30ac33cf3/pkg/apis/core/validation/validation.go#L5929) - validation missing, listMapKey=name

A second pair of eyes on this would be really helpful. I only looked at the validation code, I haven't actually tried all these cases.

**Comment by apelisse** [comment]
I wouldn't be surprised if some people think this will automatically write the validation for them. For most people authoring new resources and using this, it should tbh. Maybe we should generate and have an opt-out flag?

**Comment by jpbetz** [comment]
> Maybe we should generate and have an opt-out flag?

https://github.com/kubernetes/enhancements/pull/3938 proposes that we automatically validate. An opt-out flag sounds useful.

For these 4 ports cases, should we chnage listMapKey to 'name' to match what is validated? Can SSA handle a map key change?

**Comment by apelisse** [comment]
> For these 4 ports cases

Which 4?

If you mean:
> [Container.Ports](https://github.com/kubernetes/kubernetes/blob/5a5ebfd88b2899f2116dc858100832f30ac33cf3/pkg/apis/core/validation/validation.go#L3305C2-L3319) - validates unique name, listMapKey=containerPort, protocol

I don't think that's correct, name is unique but I don't know if that's good enough, name isn't required.

**Comment by jpbetz** [comment]
> I don't think that's correct, name is unique but I don't know if that's good enough, name isn't required.

Ah, we don't have a key then, because name is the only field that validation checks when looking for duplicates.

We might need to go with atomic?

**Comment by apelisse** [comment]
Actually, if we're careful we might be able to make it work with name, especially if it's omitempty, we can default it to "". Then as you mentioned, we would never have two keys with the same name.

**Comment by apelisse** [comment]
If that works by the way, that might be a great way to fix a lot of these problems. For env, we might be able to make a key with name AND value. Then we can only have one such pair, I think that's fine, we can still merge the list properly.

**Comment by jpbetz** [comment]
> If that works by the way, that might be a great way to fix a lot of these problems. For env, we might be able to make a key with name AND value. Then we can only have one such pair, I think that's fine, we can still merge the list properly.

I've love to see a test suite trying out various operations on a listType=map defined this way.  For CEL in CRDs we defined map key to be the CEL expression and it works just fine...

**Comment by jpbetz** [comment]
> We can reason and see what happens (and maybe fix) when we change the keys, I think we had thought about this a while back.

Good point. 

Maybe open a PR that tries to fix the broken cases and we can manually try various apply requests out?

- Env: Use "key", "value"?  This is still not exactly what we validate-- right now you could repeat `A=x{$A}` twice... should we go this route or just mark it as atomic?
- Ports: Use enough fields to ensure uniqueness in each case?  E.g. all the fields of ContainerPort?
- Cases with no validation: Mark them as atomic?

ObjectMeta.OwnerReferences  worries me..

**Comment by thockin** [comment]
> Actually, if we're careful we might be able to make it work with name, especially if it's omitempty, we can default it to "". Then as you mentioned, we would never have two keys with the same name.

For ContainerPort, the name is optional.  If specified, it must be unique, but many ports may be named "".

**Comment by thockin** [comment]
> Ports: Use enough fields to ensure uniqueness in each case? E.g. all the fields of ContainerPort?

Sadly, we allow total duplicates today, as long as `hostPort` and `hostIP` are both unspecified.  It's meaningless, but allowed.

**Comment by apelisse** [comment]
To be clear, most lists should not be atomic. Atomic should be used for things that can't be a mix of multiple things, or that are very tightly coupled, kind of like a number (wouldn't make sense to let different people change different digits), or a string (we should never let different actors change different parts of a string). 

We should try really really hard to not make these lists atomic and give them a good way to be updated by multiple actors with different opinions. Here is a litmus test: if two actors (users and/or controllers) can have different opinions of a list that are not mutually exclusive, then the list shouldn't be atomic.

Actually, atomic is breaking Ci/CD systems so much that I'd rather find a work-around to make things not completely break than make them atomic.

**Comment by apelisse** [comment]
> > LabelSelectorRequirement.Key - validation missing, patchMergeKey=key
> 
> This field is a string - what does it even mean to have a merge key?

That has to be a bug 😂

**Comment by jpbetz** [comment]
For StorageVersion: https://github.com/kubernetes/kubernetes/pull/118520

**Comment by jpbetz** [comment]
> That has to be a bug 😂 

It is. It got added by https://github.com/kubernetes/kubernetes/pull/44121 but then got missed when the bad merge keys were cleaned up by https://github.com/kubernetes/kubernetes/pull/50257.

Here's a fix: https://github.com/kubernetes/kubernetes/pull/118522

**Comment by thockin** [comment]
Antoine, I hear what you are saying but look at these fields. I think most
of them are exceedingly unlikely to be co-owned, but I admit some might be.

For some, like ContainerPort, we could add all the fields, if only SSA
would not barf on it (but could dedup it).  I don't know if that would work
on any others.

Env, as a pathological example, is just NOT a map.  It's not.  It is
sequential. We would need some sort of append-merge.

**Comment by thockin** [comment]
xref: https://github.com/kubernetes/kubernetes/pull/118475

**Comment by jpbetz** [comment]
I've [verified](https://gist.github.com/jpbetz/c0d4b0c80b846bd7b66b9a03269ada53) that all remaining cases (env, ports, imagePullSecrets, hostAliases, serviceAccount.Secrets, ownerReferences) allow exact duplicates.  This means these none of these cases cannot be fixed solely by changing the merge keys.

This simplifies things in my mind because it means we either have to handle exact duplicates in SSA, or we have to use atomic.

It is interesting that `ownerReferences` is deprecating duplicates. The warning it returns for duplicates is:

```sh
Warning: .metadata.ownerReferences contains duplicate entries; API server dedups owner references in 1.20+, and may reject such requests as early as 1.24; please fix your requests; duplicate UID(s) observed: 421f3acc-5bfd-4007-b28a-4b0749c95888
```

`ports` and `serviceAccount.secrets` have no warnings. We should add those.

Should we also consider if some fields should have a warning that we will stop allowing exact duplicates in the future?

**Comment by thockin** [comment]
I am skeptical that we can or should ever *reject* duplicates

**Comment by jpbetz** [comment]
> I am skeptical that we can or should ever _reject_ duplicates, but yes to warnings.

SGTM. When adding the warnings for the fields that do not yet have warnings, we could also update the warning for `ownerReferences` to stop saying that we might start rejecting requests in 1.24 (since it's over a year ago anyway).

**Comment by rmohr** [comment]
From the discussion above it is not entirely clear on what you will consider valid on ports. Will duplicate portnumbers with different port names still be considered valid? They are useful to aggregate pods to different services for different functions without having to expose a different port.

**Comment by thockin** [comment]
That depends on the operation.  `create` and `replace` will preserve pod-ports with the same number but different name.  `apply` will merge them based on port number+protocol (which is how the merge-key is defined)

**Comment by apelisse** [comment]
Fixed by #254
